File: src/typing/polarity/polarity.ml
(**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type t =
  | Negative
  | Neutral
  | Positive

(* Subtype relation for polarities, interpreting neutral as positive &
   negative: whenever compat(p1,p2) holds, things that have polarity p1 can
   appear in positions that have polarity p2. *)
let compat = function
  | (Positive, Positive)
  | (Negative, Negative)
  | (Neutral, _) ->
    true
  | _ -> false

let inv = function
  | Positive -> Negative
  | Negative -> Positive
  | Neutral -> Neutral

let mult = function
  | (Positive, Positive) -> Positive
  | (Negative, Negative) -> Positive
  | (Neutral, _)
  | (_, Neutral) ->
    Neutral
  | _ -> Negative

(* printer *)
let string = function
  | Positive -> "covariant"
  | Negative -> "contravariant"
  | Neutral -> "invariant"

let sigil = function
  | Positive -> "+"
  | Negative -> "-"
  | Neutral -> ""

File: rust_port/crates/flow_common/src/polarity.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use dupe::Dupe;

#[derive(
    Debug,
    Clone,
    Copy,
    Dupe,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    serde::Serialize,
    serde::Deserialize
)]
pub enum Polarity {
    Negative,
    Neutral,
    Positive,
}

impl Polarity {
    // Subtype relation for polarities, interpreting neutral as positive &
    // negative: whenever compat(p1,p2) holds, things that have polarity p1 can
    // appear in positions that have polarity p2. *)
    pub fn compat(p1: Polarity, p2: Polarity) -> bool {
        match (p1, p2) {
            (Polarity::Positive, Polarity::Positive)
            | (Polarity::Negative, Polarity::Negative)
            | (Polarity::Neutral, _) => true,
            _ => false,
        }
    }

    pub fn inv(self) -> Polarity {
        match self {
            Polarity::Positive => Polarity::Negative,
            Polarity::Negative => Polarity::Positive,
            Polarity::Neutral => Polarity::Neutral,
        }
    }

    pub fn mult(p1: Polarity, p2: Polarity) -> Polarity {
        match (p1, p2) {
            (Polarity::Positive, Polarity::Positive) => Polarity::Positive,
            (Polarity::Negative, Polarity::Negative) => Polarity::Positive,
            (Polarity::Neutral, _) | (_, Polarity::Neutral) => Polarity::Neutral,
            _ => Polarity::Negative,
        }
    }

    pub fn equal(p1: Polarity, p2: Polarity) -> bool {
        match (p1, p2) {
            (Polarity::Positive, Polarity::Positive)
            | (Polarity::Negative, Polarity::Negative)
            | (Polarity::Neutral, Polarity::Neutral) => true,
            _ => false,
        }
    }

    pub fn string(self) -> &'static str {
        match self {
            Polarity::Positive => "covariant",
            Polarity::Negative => "contravariant",
            Polarity::Neutral => "invariant",
        }
    }

    pub fn sigil(self) -> &'static str {
        match self {
            Polarity::Positive => "+",
            Polarity::Negative => "-",
            Polarity::Neutral => "",
        }
    }

    pub fn object_literal_polarity(const_: bool) -> Polarity {
        if const_ {
            Polarity::Positive
        } else {
            Polarity::Neutral
        }
    }

    pub fn apply_const(const_: bool, p: Polarity) -> Polarity {
        if const_ { Polarity::Positive } else { p }
    }
}
